core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+#[metrics(label = "stage", rename_all = "snake_case")]
+pub(crate) enum InitialStage {
+    FetchMetadataFromMainNode,
+    ApplyFactoryDeps,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "snapshots_applier")]
+pub(crate) struct SnapshotsApplierMetrics {
+    /// Number of chunks in the applied snapshot. Set when snapshots applier starts.
+    pub storage_logs_chunks_count: Gauge<usize>,
+
+    /// Number of chunks left to apply.
+    pub storage_logs_chunks_left_to_process: Gauge<usize>,
+
+    /// Total latency of applying snapshot.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub snapshot_applying_duration: Histogram<Duration>,
+
+    /// Latency of initial recovery operation split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub initial_stage_duration: Family<InitialStage, Histogram<Duration>>,
+
+    /// Latency of storage log chunk processing split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub storage_logs_chunks_duration: Family<StorageLogsChunksStage, Histogram<Duration>>,
+}
+
+#[vise::register]
+pub(crate) static METRICS: vise::Global<SnapshotsApplierMetrics> = vise::Global::new();
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotsApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error(transparent)]
+    Retryable(anyhow::Error),
+}
+pub struct SnapshotsApplier<'a> {
+    connection_pool: &'a ConnectionPool,
+    blob_store: &'a dyn ObjectStore,
+    applied_snapshot_status: SnapshotRecoveryStatus,
+}
+
+impl From<ObjectStoreError> for SnapshotsApplierError {
+    fn from(error: ObjectStoreError) -> Self {
+        match error {
+            ObjectStoreError::KeyNotFound(_) | ObjectStoreError::Serialization(_) => {
+                Self::Fatal(error.into())
+            }
+            ObjectStoreError::Other(_) => Self::Retryable(error.into()),
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotsApplierError {
+    fn from(error: SqlxError) -> Self {
+        match error {
+            SqlxError::Database(_)
+            | SqlxError::RowNotFound
+            | SqlxError::ColumnNotFound(_)
+            | SqlxError::Configuration(_)
+            | SqlxError::TypeNotFound { .. } => Self::Fatal(error.into()),
+            _ => Self::Retryable(error.into()),
+        }
+    }
+}
+
+impl From<RpcError> for SnapshotsApplierError {
+    fn from(error: RpcError) -> Self {
+        match error {
+            Error::Transport(_) | Error::RequestTimeout | Error::RestartNeeded(_) => {
+                Self::Retryable(error.into())
+            }
+            _ => Self::Fatal(error.into()),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn fetch_l2_block(&self, number: MiniblockNumber) -> Result<Option<SyncBlock>, RpcError>;
```

Comment:
What's wrong with what we are currently doing, i.e. what's your biggest motivation for suggesting this change?

Reply:
It unnecessarily ties trait implementations to a particular implementation you have in mind (i.e., using `en_syncL2Block` RPC endpoint). Not saying that it's a dealbreaker, but I'd argue that such tight coupling is not necessary in this case, given that there are viable alternatives that I've listed above, and only a relatively small part of the returned `SyncBlock` is actually used in the recovery logic. I won't insist on this change; it's mostly a stylistic issue rather than a practical one.

Reply:
I get your argument, but I'm not convinced the difference in code quality would be substantial enough to justify making this change